spot_ros2_control/hardware/src/spot_hardware_interface.cpp

```diff
@@ -0,0 +1,356 @@
+                                                                     &hw_states_[interfaces_per_joint_ * i + 1]));
+    state_interfaces.emplace_back(hardware_interface::StateInterface(joint.name, hardware_interface::HW_IF_EFFORT,
+                                                                     &hw_states_[interfaces_per_joint_ * i + 2]));
+  }
+  return state_interfaces;
+}
+
+std::vector<hardware_interface::CommandInterface> SpotHardware::export_command_interfaces() {
+  std::vector<hardware_interface::CommandInterface> command_interfaces;
+  for (uint i = 0; i < info_.joints.size(); i++) {
+    const auto& joint = info_.joints.at(i);
+    command_interfaces.emplace_back(hardware_interface::CommandInterface(joint.name, hardware_interface::HW_IF_POSITION,
+                                                                         &hw_commands_[interfaces_per_joint_ * i]));
+    command_interfaces.emplace_back(hardware_interface::CommandInterface(joint.name, hardware_interface::HW_IF_VELOCITY,
+                                                                         &hw_commands_[interfaces_per_joint_ * i + 1]));
+    command_interfaces.emplace_back(hardware_interface::CommandInterface(joint.name, hardware_interface::HW_IF_EFFORT,
+                                                                         &hw_commands_[interfaces_per_joint_ * i + 2]));
+  }
+  return command_interfaces;
+}
+
+hardware_interface::CallbackReturn SpotHardware::on_activate(const rclcpp_lifecycle::State& /*previous_state*/) {
+  // command and state should be equal when starting
+  hw_commands_.assign(hw_states_.begin(), hw_states_.end());
+  // Set up the robot using the BD SDK.
+  if (!authenticate_robot(hostname_, username_, password_)) {
+    return hardware_interface::CallbackReturn::ERROR;
+  }
+  if (!start_time_sync()) {
+    return hardware_interface::CallbackReturn::ERROR;
+  }
+  if (!check_estop()) {
+    return hardware_interface::CallbackReturn::ERROR;
+  }
+  if (!get_lease()) {
+    return hardware_interface::CallbackReturn::ERROR;
+  }
+  if (!power_on()) {
+    return hardware_interface::CallbackReturn::ERROR;
+  }
+  if (!start_state_stream(std::bind(&StateStreamingHandler::handle_state_streaming, &state_streaming_handler_,
+                                    std::placeholders::_1))) {
+    return hardware_interface::CallbackReturn::ERROR;
+  }
+
+  return hardware_interface::CallbackReturn::SUCCESS;
+}
+
+hardware_interface::CallbackReturn SpotHardware::on_deactivate(const rclcpp_lifecycle::State& /*previous_state*/) {
+  stop_state_stream();
+  release_lease();
+  return hardware_interface::CallbackReturn::SUCCESS;
+}
+
+hardware_interface::CallbackReturn SpotHardware::on_shutdown(const rclcpp_lifecycle::State& /*previous_state*/) {
+  stop_state_stream();
+  release_lease();
+  return hardware_interface::CallbackReturn::SUCCESS;
+}
```

Comment:
@khughes-bdai how does this work? Can we stop the state stream and release the lease twice? I ask because that is a likely scenario looking at the FSM (see https://design.ros2.org/articles/node_lifecycle.html, e.g. activate -> deactivate -> shutdown).

Reply:
Ahh, yes I meant to fix this. I think both of these should only happen in deactivate, will update

Reply:
Follow up question @mhidalgo-bdai: Does it make more sense to do the robot authentication, acquiring the lease, starting the state stream, etc. in configure or activate? Currently it happens in activate.

Reply:
There are no strong requirements here, it's mostly convention. I would personally move authentication and time syncing to configure (if either fails, you'll probably end up having to reboot), and leave the rest to activate (a leasing failure could mean the robot is taken already), but that's me.

Reply:
I found some reasonable guidelines in [`ros2_control` sources](https://github.com/ros-controls/ros2_control/blob/5cd1a43f8c95b5586ae4d5728e5052892ae1c80e/hardware_interface/include/hardware_interface/system_interface.hpp#L49-L68).

Reply:
More reasonable structure added in latest commit -- we don't need the lease to state stream.